feat: Create Swift Package

barbeau opened the pull request:
...when running:

`./gradlew createSwiftPackage`

This is an attempt to close https://github.com/CUTR-at-USF/opentripplanner-client-library/issues/7 by using the process discussed at https://johnoreilly.dev/posts/kotlinmultiplatform-swift-package/ to generate a Swift Package.

I can't test this on my local Windows machine but let's see if we can get it to work on CI.

barbeau committed:
feat: Create Swift Package

When running:
./gradlew createSwiftPackage

barbeau committed:
ci: Add XCode setup

barbeau committed:
Fix whitespace

barbeau committed:
ci: Switch to Mac OS

barbeau commented:
@aaronbrethorst @ualch9 could you please try cloning this repo, checking out this PR branch (`ios-package`), and running `./gradlew createSwiftPackage`?

🤞  it should generate Swift package files in `common/swiftpackage` (according to [the article](https://johnoreilly.dev/posts/kotlinmultiplatform-swift-package/)).

I apparently can't run this because I'm on a Windows machine, and I've having some issues getting GitHub Actions CI to build the project when including XCode even after switching to building on `macos-latest` instead of  `ubuntu-latest` (see [this build](https://github.com/CUTR-at-USF/opentripplanner-client-library/runs/2273327432)). I was stealing lines from the [OBAKit repo GitHub Action](https://github.com/OneBusAway/OBAKit/blob/master/.github/workflows/obakittests.yml) for this PR, but let me know if you have any tips on getting CI working.

barbeau committed:
fix: Add the rest of the example from the blog

Parts of the Dispatcher.kt for iOS were missing when compared to https://www.raywenderlich.com/19144111-kotlin-multiplatform-project-for-android-and-ios-getting-started.

barbeau commented:
Success 🎉🎉!

CI now builds and creates a Swift Package (although it doesn't do anything with it yet) - https://github.com/CUTR-at-USF/opentripplanner-client-library/runs/2273496842

Looks like we were just missing part of the iOS sample code from the example. So in theory this should build locally for you fine now. I believe you will need Java 8 or higher installed on your machine to run Gradle build from the command line, but don't necessarily need Android Studio installed.

barbeau committed:
docs: Add iOS package info to README

barbeau committed:
chore: Cleanup iOS code

ualch9 commented on .github/workflows/gradle.yml:
```diff
@@ -12,15 +12,22 @@ on:
+    - name: Set up XCode
+      uses: maxim-lobanov/setup-xcode@master
+      with:
+        xcode-version: '12.2'
```
current xcode version is `12.4`. 12.4 on the ci should work because it worked on my machine.

barbeau replied:
Changed in https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12/commits/cece4885e2f7085b31b425af51e091d0df87470a

ualch9 commented:
@barbeau installing jdk 11 and running `./gradlew createSwiftPackage` works for me. However, the output framework name is `library`, when it should be `OpenTripPlannerClientLibrary`, resulting in the following usage code:
```swift
import library     // ← ! should be OpenTripPlannerClientLibrary

class FirstView: View {
   // ...
}
```

edit:
The above is directly including the built `xcframework` into Xcode (without using spm).
I also tried this using SPM since the Package.swift file uses `OpenTripPlannerClientLibrary`, but the same issue is present.

SPM: https://github.com/ualch9/opentripplanner-library-swift

barbeau commented:
@ualch9 Thanks for checking it out, I'll take a look.

So would the best practice be to publish the SPM artifact within this repo? If I'm understanding correctly that seems to be what's done in https://github.com/apple/swift-protobuf.

aaronbrethorst commented:
@barbeau I'm not certain about Swift-Protobuf, but Apple has some documentation about publishing binary artifacts within a `Package.swift` manifest here: https://developer.apple.com/documentation/swift_packages/distributing_binary_frameworks_as_swift_packages

and here's another writeup on how a 3rd party developer distributes their binary frameworks via SPM: https://pspdfkit.com/blog/2020/binary-frameworks-as-swift-packages/

ualch9 commented:
> So would the best practice be to publish the SPM artifact within this repo?

Yes. However, SPM does not have a "build" command or script, so everytime `master` is updated, you'd have to also push a new `xcframework`. Since apple/swift-protobuf is a Swift package written in Swift/C++, they don't need to provide a binary.

I was in the middle of writing this comment when aaron responded, but binary distibution in the git repo is discouraged because spm downloads the repo's history as well.

https://pspdfkit.com/blog/2020/binary-frameworks-as-swift-packages/#targets
> While it’s possible to check the binaries into the git repository, we decided to host the binaries on our own servers because Apple doesn’t recommend checking the binaries into the repository. This is to prevent git checkout slowdowns.

barbeau committed:
ci: Change to Xcode 12.4

barbeau committed:
fix: Change iOS framework name to the correct name

barbeau commented:
>However, the output framework name is library, when it should be OpenTripPlannerClientLibrary, 

Looks like the same issue as https://github.com/ge-org/multiplatform-swiftpackage/issues/25.

@ualch9 I believe I've fixed this in https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12/commits/faadf7a9f8edb407641befafe45adf0bba44428a in this branch. Could you give it another try and see if that looks better?

ualch9 commented on library/build.gradle.kts:
```diff
@@ -15,7 +15,7 @@ kotlin {
         binaries {
             framework {
-                baseName = "library"
+                baseName = "OpenTripPlannerClientLibrary"
```
👍

ualch9 commented:
Looks good, the following iOS Swift code compiles:
```swift
import OpenTripPlannerClientLibrary

let rentalAPI = BikeRentalApi(
    url: "https://10.0.2.2:8080/otp/routers/default/bike_rental", 
    locale: "en_US",                // ← I'm not sure if the value I provided is valid.
    lowerLeft: "41.81712,-87.62678", 
    upperRight: "41.84584,-87.65214")

rentalAPI.getBikeRental { (response) in
    for station in response.stations! {
        print(station)
    }
} failure: { (error) in
    print("Error: \(error?.message)")
}
```

macOS Swift should also be supported given that Kotlin desktop is a target for this project. Although worth mentioning that it is possible to run iOS code on macOS using Mac Catalyst or natively on an M1.

barbeau commented:
To fix the iOS code in the master branch, let's go ahead and finish reviewing/merging this PR, and I'll work on publishing the artifacts separately (see issue https://github.com/CUTR-at-USF/opentripplanner-client-library/issues/14).

@BumbleFlash it should be ready for review.

barbeau commented:
>I was in the middle of writing this comment when aaron responded, but binary distibution in the git repo is discouraged because spm downloads the repo's history as well.

@ualch9 Thanks, I was concerned about this too. I'll look at other options for hosting the artifacts in a separate issue/PR.

barbeau committed:
Merge branch 'main' into ios-package

barbeau commented:
Hmm, when merging the main branch that includes the JVM build (#13), CI is now failing:
https://github.com/CUTR-at-USF/opentripplanner-client-library/runs/2298307398?check_suite_focus=true

with:

```
...
[Step 1 of 1] Preparing commonized Kotlin/Native libraries for targets [ios_x64], [ios_arm64] (280 items): Done


e: Could not find "kotlinx-coroutines-core_concurrentMain" in [/Users/runner/work/opentripplanner-client-library/opentripplanner-client-library, /Users/runner/.konan/klib, /Users/runner/.konan/kotlin-native-prebuilt-macos-1.4.32/klib/common, /Users/runner/.konan/kotlin-native-prebuilt-macos-1.4.32/klib/platform/ios_arm64].
Daemon vm is shutting down... The daemon has exited normally or was terminated in response to a user interrupt.
----- End of the daemon log -----


FAILURE: Build failed with an exception.

* What went wrong:
Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

> Task :library:compileIosMainKotlinMetadata
e: Could not find "kotlinx-coroutines-core_concurrentMain" in [/Users/runner/work/opentripplanner-client-library/opentripplanner-client-library, /Users/runner/.konan/klib, /Users/runner/.konan/kotlin-native-prebuilt-macos-1.4.32/klib/common, /Users/runner/.konan/kotlin-native-prebuilt-macos-1.4.32/klib/platform/ios_arm64].
Error: Process completed with exit code 1.
```

barbeau committed:
chore: Use variable for coroutine version

barbeau committed:
chore: Remove maven local reference in library

barbeau committed:
chore: Bump ktor and coroutine versions

barbeau commented:
Looks like this may be the issue:
https://youtrack.jetbrains.com/issue/KT-43651

BumbleFlash commented:
@barbeau could you try adding this to the `iosMain` module in the multiplatform gradle file?
```
implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.4.2-native-mt"
```

if this doesn't work, I think you can use `strictly` keyword to force it to use the multithreaded version. This link should help: https://kotlinlang.org/docs/mobile/concurrency-and-coroutines.html#multithreaded-coroutines

barbeau committed:
chore: Remove JVM co-routine dependency

May be triggering https://youtrack.jetbrains.com/issue/KT-43651

barbeau commented on library/build.gradle.kts:
```diff
@@ -1,22 +1,19 @@
 group = "edu.usf.cutr.otp"
 version = "1.0.0"
 
-repositories {
```
@BumbleFlash See https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/13/files#r609862320. Can we get rid of this?

BumbleFlash replied:
Yes, I think you can. I'm not getting any local dependencies so I don't think we need it.

barbeau commented on library/build.gradle.kts:
```diff
@@ -60,7 +57,6 @@ kotlin {
         val jvmMain by getting {
             dependencies {
                 implementation("io.ktor:ktor-client-java:$ktorVersion")
-                implementation ("org.jetbrains.kotlinx:kotlinx-coroutines-swing:1.4.2")
```
@BumbleFlash I think this might be what's triggering https://youtrack.jetbrains.com/issue/KT-43651, so I'm going to try removing it. The JVM project still seems to build fine without it.

barbeau replied:
Hmm, no luck, same error at https://github.com/CUTR-at-USF/opentripplanner-client-library/runs/2298725831.

BumbleFlash replied:
In that case, I guess you can. I got a runtime error when I included the `core` library instead of the `swing`. Did you try running to see if it stills logs data if you haven't tried already?

barbeau replied:
>I got a runtime error when I included the core library instead of the swing. Did you try running to see if it stills logs data if you haven't tried already?

You mean running with the test desktop JVM project? No, I haven't tried - could you try that to see if we need this dependency?

I opened a test PR at https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/15 to see if this is an issue when building on macOS, just to see if the existing main branch builds on macOS.

BumbleFlash replied:
Just tested the code, the desktop throws a run time error if we don't include the `swing` coroutine dependency. I can add that back to this PR. 

The dependency provides the main dispatcher to the library. It acts quite similar to the `coroutine-android` library in the `androidMain` module.

barbeau replied:
Thanks for testing, I just re-added it in https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12/commits/14053ccc540e9e3eac9cac1189c779f34cfa34ed

barbeau committed:
chore: Re-add JVM coroutine dependency

Otherwise we get runtime error in client application - see https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12#discussion_r609902863

BumbleFlash commented:
@barbeau Looks like adding the native dependency mentioned above https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12#issuecomment-815961091 fixes the build error in #15. Could you try adding it to this PR and see if it builds?

barbeau committed:
fix: Add multithreaded dependency on iOS

See https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12#issuecomment-816002015

barbeau commented:
@BumbleFlash Thanks for testing! I pushed that change in https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12/commits/29e5f261a73ac69c6e64619308930d8c23c41d4a.

Here's my understanding of what's happening - io.ktor:ktor-client-ios, which we're using in the iOS library, has a dependency on kotlinx-coroutines-core:1.4.3-native-mt. This worked fine originally in this PR because the JVM build wasn't included. When we included the JVM build, io.ktor:ktor-client-java dependency for JVM started overriding the kotlinx-coroutines-core:1.4.3-native-mt dependency with kotlinx-coroutines-core:1.4.3, which caused the build failure on CI.

barbeau committed:
fix: Use variable for coroutine version

barbeau commented:
Hmmm, still no luck.

BumbleFlash commented:
Hmmm. something's not right. Could you try adding `strictly `keyword to force it to use the multithreaded version in the common module? This link should help https://kotlinlang.org/docs/mobile/concurrency-and-coroutines.html#multithreaded-coroutines.

I'll test it on my local machine to see if it affects the JVM build.

BumbleFlash commented:
Actually, could you try adding that to the `iosModule`? 

> Because the main version of kotlinx.coroutines is a single-threaded one, libraries will almost certainly rely on this version

Looks like `iosMain` is still relying on the core dependency.

barbeau committed:
fix: Use strictly to prevent replacement of version

BumbleFlash commented:
Could you try replacing the `coroutine-core` in the commonMain with the `native` one excluding the `strictly `keyword? I tested it on my local machine and it doesn't seem to affect the JVM Build.

barbeau committed:
fix: Add multithreaded dependency on common

See https://github.com/CUTR-at-USF/opentripplanner-client-library/pull/12#issuecomment-816002015

BumbleFlash commented:
Great job! Thank you! Looks like `native` in the common module does the trick. I'll test it on my machine once again since it included the `strictly` keyword

BumbleFlash reviewed (approved):
LGTM! I tested it on my local machine and the JVM build still runs fine. I guess we can go ahead and merged this.
